src/bless-boot/bless-boot.c

```diff
@@ -236,23 +234,32 @@ static int acquire_boot_count_path(
                                        "Path read from LoaderBootCountPath is not absolute, refusing: %s",
                                        path);
 
-        last = last_path_component(path);
-        e = strrchr(last, '+');
+        _cleanup_free_ char *last = NULL;
+        r = path_extract_filename(path, &last);
+        if (r < 0)
+                return log_error_errno(r, "Failed to extract filename from LoaderBootCountPath '%s': %m", path);
+        if (r == O_DIRECTORY)
+                return log_error_errno(SYNTHETIC_ERRNO(EISDIR), "LoaderBootCountPath '%s' refers to directory path, refusing.", path);
+
+        const char *e = strrchr(last, '+');
```

Comment:
`last` is a new string (`a = strndup(c, r); *ret = TAKE_PTR(a);` from `int path_extract_filename(const char *path, char **ret)`).

Here we get a pointer to `+` character in `last`, but then `prefix = strndup(path, e - path);` does pointer math between two different strings, to `strndup` will always copy the entire `path` into `prefix`.

Reply:
ouch